NEXT-23421 - Tax provider documentation (#798)
shopware/docs: guides/plugins/plugins/checkout/cart/tax-provider.md

# Tax provider

## Overview

Tax calculations differ from country to country. Especially in the US, the sales tax calculation can be tedious, as the laws and regulation differ from state to state or even county or certain city taxes can apply. Therefore, most shops use a third-party service (so called tax-providers) to calculate sales taxes for them.

Starting with version 6.5.0.0, Shopware allows plugins to integrate custom tax calculation, which could include an automatic tax calculation with a tax provider. A plugin should provide a class extending the `Shopware\Core\Checkout\Cart\TaxProvider\AbstractTaxProvider`, which is called during checkout and provide new tax rates.

## Prerequisites

As most guides, this guide is also built upon our [plugin base guide](../../plugin-base-guide.md), but it's not mandatory to use exactly that plugin as a foundation. The examples in this guide use the namespace however.

## Creating a tax provider

> First of all, we need to create a class which handles the tax calculation or calls you preferred tax provider.

Reviewer: youR tax provider

> For an easy start, extend you class from `Shopware\Core\Checkout\Cart\TaxProvider\AbstractTaxProvider` and implement the `provideTax` method.

Reviewer: youR class

> You probably want to call a tax provider, who will calculate the taxes for you. In our case, we simply apply a hefty 50% tax rate for all line-items in the cart.

Reviewer: ~~who~~ which

    {
        $lineItemTaxes = [];
    
        foreach ($cart->getLineItems() as $lineItem) {
            $taxRate = 50;
            $price = $lineItem->getPrice()->getTotalPrice();
            $tax = $price * $taxRate / 100;
        
            $lineItemTaxes[$lineItem->getId()] = new CalculatedTaxCollection(
                [
                    new CalculatedTax($tax, $taxRate, $price),
                ]
            );
        }
        
        // you could do the same for deliveries
        // $deliveryTaxes = [];
        
        // foreach ($cart->getDeliveries() as $delivery) {
        // ...
        // }
        
        // If you call a tax provider, you will probably get calculated tax sums for the whole cart
        // Use the cartPriceTaxes to let Shopware show the correct sums in the checkout
        // If omitted, Shopware will try to calculate the tax sums itself
        // $cartPriceTaxes = [];
        
        return new TaxProviderStruct(
            $lineItemTaxes,
            // $deliveries,
            // $cartPriceTaxes
        );
    }
}

```

{% endcode %}

## Registering the tax provider in the DI container

> If you've created your tax provider, you need to register it in the DI container. To do so, you need to create a new service in the `services.xml` file and tag the service as `shopware.tax_provider`.

Reviewer: ~~If~~ When

{% code title="<plugin root>/src/Resources/config/services.xml" %}

```markup
<?xml version="1.0" ?>

<container xmlns="http://symfony.com/schema/dic/services"
           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">

    <services>
        <service id="SwagTaxProviders\TaxJar\Provider\TaxJarProvider">
            <tag name="shopware.tax.provider" />
        </service>
    </services>
```

{% endcode %}

## Migrate your tax provider to the database

> Last but not least, you will have to create a migration to add your tax provider to the database. This is necessary, so that Shopware knows of your tax provider in the first place during checkout.

Reviewer: You could create it via repository during the Plugin installation process as well, right? Maybe you can write a sentence, that this is a valid alternative?

            'id' => $id,
            'identifier' => Swag\BasicExample\Checkout\Cart\Tax\TaxProvider::class, // your tax provider class
            'active' => true,
            'priority' => 1,
            'availability_rule_id' => $ruleId,
            'created_at' => (new \DateTime())->format(Defaults::STORAGE_DATE_TIME_FORMAT),
        ]);
    }

    public function updateDestructive(Connection $connection): void
    {
    }
}
```

{% endcode %}

### Via repository

If you don't want to use migrations, you can also create your tax provider via repository. A good place for doing so would be the `install` lifecycle method of your plugin.

If you do not know of plugin lifecycle methods yet, please refer to our [plugin lifecycle guide](../../plugin-fundamentals/plugin-lifecycle.md).

{% code title="<plugin root>/src/BasicExample.php" %}

```php
<?php declare(strict_types=1);

namespace SwagBasicExample;

use Shopware\Core\Framework\DataAbstractionLayer\Search\Criteria;
use Shopware\Core\Framework\DataAbstractionLayer\Search\Filter\EqualsFilter;
use Shopware\Core\Framework\Plugin;
use Shopware\Core\Framework\Plugin\Context\InstallContext;
use Shopware\Core\Framework\Uuid\Uuid;

> if (\file_exists(\dirname(__DIR__) . '/vendor/autoload.php')) {

Reviewer 2: Why is it necessary to load dependencies here? I do not see it being used in the example. As they are loaded in 6.5 anyways I think this it is better to remove it as it adds more cognitive load to understand the example.

Author: Thank you Joshua, you are completely right, this is a copy paste issue ;)

        ]);
    }

    public function updateDestructive(Connection $connection): void
    {
    }
}
```

{% endcode %}

### Via repository

If you don't want to use migrations, you can also create your tax provider via repository. A good place for doing so would be the `install` lifecycle method of your plugin.

If you do not know of plugin lifecycle methods yet, please refer to our [plugin lifecycle guide](../../plugin-fundamentals/plugin-lifecycle.md).

{% code title="<plugin root>/src/BasicExample.php" %}

```php
<?php declare(strict_types=1);

namespace SwagBasicExample;

use Shopware\Core\Framework\DataAbstractionLayer\Search\Criteria;
use Shopware\Core\Framework\DataAbstractionLayer\Search\Filter\EqualsFilter;
use Shopware\Core\Framework\Plugin;
use Shopware\Core\Framework\Plugin\Context\InstallContext;
use Shopware\Core\Framework\Uuid\Uuid;

class SwagBasicExample extends Plugin
{
    public function install(InstallContext $installContext): void
    {
        $ruleRepo = $this->container->get('rule.repository');

        $criteria = new Criteria();
        $criteria->addFilter(new EqualsFilter('name', 'Always valid (Default)'));

>         $rule = $ruleRepo->search($criteria, $installContext->getContext())->first();

Reviewer 2: I would not trust this to work. This will be the example for everyone copying the code and I know that we have multiple projects without this rule name. Maybe rather suggest to add a new rule with the same structure. I'd rather have a rule, that I can just edit and only influence my tax provider, than added a new rule and have to assign it later to the tax provider. Now I question myself how this is solved for apps.

        $taxRepo = $this->container->get('tax_provider.repository');
        $taxRepo->create([
            [
                'id' => Uuid::randomHex(),
                'identifier' => Swag\BasicExample\Checkout\Cart\Tax\TaxProvider::class,
                'priority' => 1,
                'active' => false, // activate this via the `activate` lifecycle method
                'availabilityRule' => $rule,
                'translations' => [
                    'de-DE' => [
                        'name' => 'TaxJar',
                    ],
                    'en-GB' => [
                        'name' => 'TaxJar',
                    ],
                ],
            ],
        ], $installContext->getContext());
    }
    
    // do not forget to add an `uninstall` method, which removes your tax providers, when the plugin is uninstalled
    // you also may want to add the `activate` and `deactivate` methods to activate and deactivate the tax providers
}

```

{% endcode %}

> If you have taken these steps, your tax provider should be ready and already used during checkout. You should also see the new tax provider showing up in the administration in `Settings > Tax`.

Reviewer 2: I'd like to see that hint in the instructions for the app as well. They wouldn't know where to see and configure these without looking here.

# Tax provider

## Overview

Tax calculations differ from country to country. Especially in the US, the sales tax calculation can be tedious, as the laws and regulation differ from state to state or even county or certain city taxes can apply. Therefore, most shops use a third-party service (so called tax providers) to calculate sales taxes for them.

Starting with version 6.5.0.0, Shopware allows plugins to integrate custom tax calculation, which could include an automatic tax calculation with a tax provider. A plugin should provide a class extending the `Shopware\Core\Checkout\Cart\TaxProvider\AbstractTaxProvider`, which is called during checkout and provide new tax rates.

## Prerequisites

> As most guides, this guide is also built upon our [plugin base guide](../../plugin-base-guide.md), but it's not mandatory to use exactly that plugin as a foundation. The examples in this guide use the namespace however.

Reviewer 3: ```suggestion
Refer to [plugin base guide](../../plugin-base-guide.md). It is not mandatory to use exactly the same plugin as a foundation.
```

## Creating a tax provider

> First of all, we need to create a class which handles the tax calculation or calls your preferred tax provider.
> 
> For an easy start, extend your class from `Shopware\Core\Checkout\Cart\TaxProvider\AbstractTaxProvider` and implement the `provide` method.
> 
> You probably want to call a tax provider, which will calculate the taxes for you. In our case, we simply apply a hefty 50% tax rate for all line-items in the cart.

Reviewer 3: ```suggestion
Firstly you need to create a class which handles the tax calculation or calls your preferred tax provider. For an easy start, extend your class from `Shopware\Core\Checkout\Cart\TaxProvider\AbstractTaxProvider` and implement the `provide` method.

You may then call a tax provider, which will calculate the taxes for you. For example, we simply apply a hefty 50% tax rate for all line-items in the cart.
```

                [
                    new CalculatedTax($tax, $taxRate, $price),
                ]
            );
        }
        
        // you could do the same for deliveries
        // $deliveryTaxes = []; // use the id of the delivery position as keys, if you want to transmit delivery taxes
        
        // foreach ($cart->getDeliveries() as $delivery) {
        //     foreach ($delivery->getPositions() as $position) {
        //         $deliveryTaxes[$delivery->getId()] = new CalculatedTaxCollection(...);
        //         ...
        //     }
        // }
        
        // If you call a tax provider, you will probably get calculated tax sums for the whole cart
        // Use the cartPriceTaxes to let Shopware show the correct sums in the checkout
        // If omitted, Shopware will try to calculate the tax sums itself
        // $cartPriceTaxes = [];
        
        return new TaxProviderResult(
            $lineItemTaxes,
            // $deliveryTaxes,
            // $cartPriceTaxes
        );
    }
}

```

{% endcode %}

## Registering the tax provider in the DI container

> After you've created your tax provider, you need to register it in the DI container. To do so, you need to create a new service in the `services.xml` file and tag the service as `shopware.tax.provider`.

Reviewer 3: ```suggestion
After you have created your tax provider, you need to register it in the DI container. To do so, you need to create a new service in the `services.xml` file and tag the service as `shopware.tax.provider`.
```

{% code title="<plugin root>/src/Resources/config/services.xml" %}

```markup
<?xml version="1.0" ?>

<container xmlns="http://symfony.com/schema/dic/services"
           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">

    <services>
        <service id="Swag\BasicExample\Checkout\Cart\Tax\TaxProvider">
            <tag name="shopware.tax.provider" />
        </service>
    </services>
```

{% endcode %}

## Migrate your tax provider to the database

> To let Shopware know of your new tax provider, you will have to persist it to the database.
> 
> You can either to do so via a migration or the entity repository.

Reviewer 3: ```suggestion
To let Shopware know of your new tax provider, you will have to persist it to the database. You can do so either via a migration or the entity repository.
```

use Shopware\Core\Framework\Migration\MigrationStep;
use Shopware\Core\Framework\Uuid\Uuid;
use Shopware\Core\System\TaxProvider\Aggregate\TaxProviderTranslation\TaxProviderTranslationDefinition;
use Shopware\Core\System\TaxProvider\TaxProviderDefinition;

class MigrationTaxProvider extends MigrationStep
{
    public function getCreationTimestamp(): int
    {
        return 1662366151;
    }

    public function update(Connection $connection): void
    {
        $ruleId = $connection->fetchOne(
            'SELECT `id` FROM `rule` WHERE `name` = :name',
            ['name' => 'Always valid (Default)']
        );

        $id = Uuid::randomBytes();

        $connection->insert(TaxProviderDefinition::ENTITY_NAME, [
            'id' => $id,
            'identifier' => Swag\BasicExample\Checkout\Cart\Tax\TaxProvider::class, // your tax provider class
            'active' => true,
            'priority' => 1,
            'availability_rule_id' => $ruleId,
            'created_at' => (new \DateTime())->format(Defaults::STORAGE_DATE_TIME_FORMAT),
        ]);
    }

    public function updateDestructive(Connection $connection): void
    {
    }
}
```

{% endcode %}

### Via repository

> You may want to have a look at the [repository guide](../../../../../concepts/framework/data-abstraction-layer#crud-operations) to learn more about on how to use the entity repository to manipulate data.

Reviewer 3: ```suggestion
You may want to have a look at the [repository guide](../../../../../concepts/framework/data-abstraction-layer#crud-operations) to learn more on how to use the entity repository to manipulate data.
```

                        'operator' => '>=',
                        'value' => 0,
                    ],
                ],
            ],
        ], $installContext->getContext());
        
        $criteria = new Criteria();
        $criteria->addFilter(
            new EqualsFilter('name', 'Cart > 0')
        );

        $ruleId = $ruleRepo->searchIds($criteria, $installContext->getContext())->firstId();

        $taxRepo = $this->container->get('tax_provider.repository');
        $taxRepo->create([
            [
                'id' => Uuid::randomHex(),
                'identifier' => Swag\BasicExample\Checkout\Cart\Tax\TaxProvider::class,
                'priority' => 1,
                'active' => false, // activate this via the `activate` lifecycle method
                'availabilityRuleId' => $ruleId,
            ],
        ], $installContext->getContext());
    }
    
    // do not forget to add an `uninstall` method, which removes your tax providers, when the plugin is uninstalled
    // you also may want to add the `activate` and `deactivate` methods to activate and deactivate the tax providers
}

```

{% endcode %}

> If you have taken these steps, your tax provider should be ready. You should see the new tax provider showing up in the administration in `Settings > Tax`, where you can change the active state, priority and availability rule manually afterwards.

Reviewer 3: ```suggestion
You should now see the new tax provider showing up in the Administration in `Settings > Tax`, where you can change the active state, priority and availability rule manually.
```
